builtin/logical/database/path_roles.go

```diff
@@ -2,263 +2,606 @@ package database
+		b.credRotationQueue.PopByKey(name)
+	}
+
+	return nil, nil
+}
+
+func (b *databaseBackend) pathStaticRoleRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	role, err := b.StaticRole(ctx, req.Storage, d.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+	data := pathRoleReadCommon(role)
+	if role.StaticAccount != nil {
+		data["username"] = role.StaticAccount.Username
+		data["rotation_period"] = role.StaticAccount.RotationPeriod.Seconds()
+		if !role.StaticAccount.LastVaultRotation.IsZero() {
+			data["last_vault_rotation"] = role.StaticAccount.LastVaultRotation
 		}
-		if len(role.Statements.Rollback) == 0 {
-			data["rollback_statements"] = []string{}
+	}
+
+	return &logical.Response{
+		Data: data,
+	}, nil
+}
+
+func (b *databaseBackend) pathRoleRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(ctx, req.Storage, d.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: pathRoleReadCommon(role),
+	}, nil
+}
+
+func pathRoleReadCommon(role *roleEntry) map[string]interface{} {
+	data := map[string]interface{}{
+		"db_name":               role.DBName,
+		"creation_statements":   role.Statements.Creation,
+		"revocation_statements": role.Statements.Revocation,
+		"rollback_statements":   role.Statements.Rollback,
+		"renew_statements":      role.Statements.Renewal,
+		"rotation_statements":   role.Statements.Rotation,
+		"default_ttl":           role.DefaultTTL.Seconds(),
+		"max_ttl":               role.MaxTTL.Seconds(),
+	}
+	if len(role.Statements.Creation) == 0 {
+		data["creation_statements"] = []string{}
+	}
+	if len(role.Statements.Revocation) == 0 {
+		data["revocation_statements"] = []string{}
+	}
+	if len(role.Statements.Rollback) == 0 {
+		data["rollback_statements"] = []string{}
+	}
+	if len(role.Statements.Renewal) == 0 {
+		data["renew_statements"] = []string{}
+	}
+	if len(role.Statements.Rotation) == 0 {
+		data["rotation_statements"] = []string{}
+	}
+	return data
+}
+
+func (b *databaseBackend) pathRoleList(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	path := databaseRolePath
+	if req.Path == "static-roles/" {
+		path = databaseStaticRolePath
+	}
+	entries, err := req.Storage.List(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *databaseBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	if name == "" {
+		return logical.ErrorResponse("empty role name attribute given"), nil
+	}
+
+	role, err := b.Role(ctx, req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		role = &roleEntry{}
+	}
```

Comment:
Since databaseConfig.AllowedRoles uses role name as the identifier i think it's important to not allow a role and a staticrole to share names. I think here and below we should verify there isn't already a role of the other type with the same name

Reply:
I added a unique name check and test in e1ffe3361